Automatically assign newly created flake issues to test owners.

rmmh opened the pull request.

Commit:
Automatically assign newly created flake issues based on an input list of test owners.

lavalamp commented on mungegithub/mungers/sync/issue-sync.go:
```diff
@@ -60,14 +61,16 @@ type IssueSyncer struct {
 
 // NewIssueSyncer constructs an issue syncer.
-func NewIssueSyncer(config *github.Config, finder IssueFinder) *IssueSyncer {
+func NewIssueSyncer(config *github.Config, finder IssueFinder, owner testowner.Mapper) *IssueSyncer {
```
Define an interface here and then don't import the testowner package.

lavalamp commented on mungegithub/mungers/flake-manager.go:
```diff
@@ -76,7 +79,19 @@ func (p *FlakeManager) Initialize(config *github.Config, features *features.Feat
+
+	var owner testowner.Mapper
+	if p.ownerPath != "" {
+		tsv, err := os.Open(p.ownerPath)
```
You never close this? Also what does tsv stand for?

lavalamp commented on mungegithub/mungers/testowner/owner.go:
```diff
@@ -0,0 +1,73 @@
+var whiteSpaceRegex = regexp.MustCompile(`\s+`)
+
+// Turn a test name into a canonical form (without tags, lowercase, etc.)
+func canonize(name string) string {
```
It's "canonicalize"-- canonize means something else :)

lavalamp commented on mungegithub/github/github.go:
```diff
@@ -497,16 +497,17 @@ func (config *Config) GetObject(num int) (*MungeObject, error) {
 
 // NewIssue will file a new issue and return an object for it.
-func (config *Config) NewIssue(title, body string, labels []string) (*MungeObject, error) {
+func (config *Config) NewIssue(title, body string, labels []string, owner string) (*MungeObject, error) {
```
what happens if owner is ""? How can you indicate no known owner?

lavalamp commented on mungegithub/mungers/flake-manager.go:
```diff
@@ -96,7 +111,9 @@ func (p *FlakeManager) EachLoop() error {
 // AddFlags will add any request flags to the cobra `cmd`
-func (p *FlakeManager) AddFlags(cmd *cobra.Command, config *github.Config) {}
+func (p *FlakeManager) AddFlags(cmd *cobra.Command, config *github.Config) {
+	cmd.Flags().StringVar(&p.ownerPath, "test-owners-config", "", "file containing a CSV-exported test-owners spreadsheet")
```
This is OK as a first step but we really need to get this from the repository. We can't restart the submit queue every time this changes.

@eparis might have ideas

lavalamp commented on mungegithub/mungers/sync/issue-sync.go:
```diff
@@ -201,10 +204,16 @@ func (s *IssueSyncer) createIssue(source IssueSource) (issueNumber int, err erro
 	}
 
+	owner := ""
+	if s.owner != nil {
```
Suggest never letting this be nil? Or document that it can be nil in the struct.

lavalamp commented on mungegithub/mungers/testowner/owner.go:
```diff
@@ -0,0 +1,73 @@
+
+type Mapper interface {
+	// Given a test name, provide a Github username for the test
+	Get(name string) string
```
Suggest `TestOwner(testName string) string`

lavalamp commented on mungegithub/mungers/testowner/owner.go:
```diff
@@ -0,0 +1,73 @@
+	return strings.ToLower(strings.TrimSpace(string(squeezed)))
+}
+
+type Mapper interface {
```
godoc

lavalamp commented on mungegithub/mungers/testowner/owner.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+// Construct an OwnerList given a CSV file with 'owner' and 'test name' columns.
+func NewOwnerListFromTsv(r io.Reader) (*OwnerList, error) {
```
s/TSV/CSV/

rmmh commented:
Thanks, I think I fixed everything.

I added a mapper that will reload its configuration without restarting the daemon. Ideally a small container in the submit-queue pod will generate the owners list and write it to a shared volume, but it can be done manually for now.

lavalamp commented on mungegithub/github/github.go:
```diff
@@ -503,11 +503,15 @@ func (config *Config) NewIssue(title, body string, labels []string, owner string
 	if config.DryRun {
 		return nil, fmt.Errorf("can't make issues in dry-run mode")
 	}
+	var assignee *string
```
Document this behavior in the function comment.

lavalamp commented on mungegithub/mungers/sync/issue-sync.go:
```diff
@@ -61,11 +66,11 @@ type IssueSyncer struct {
+	owner  OwnerMapper // If nil, automatic issue assignment is disabled.
 }
 
 // NewIssueSyncer constructs an issue syncer.
```
Add: "'owner' may be nil."

lavalamp commented on mungegithub/mungers/testowner/owner_test.go:
```diff
@@ -1,8 +1,12 @@
+	"time"
 )
 
 func TestCanonize(t *testing.T) {
```
TestNormalize

lavalamp commented on mungegithub/mungers/testowner/owner_test.go:
```diff
@@ -13,38 +17,82 @@ func TestCanonize(t *testing.T) {
+
+	// Assuming millisecond resolution on our FS, this sleep
+	// ensures the mtime will change with the next write.
+	time.Sleep(5 * time.Millisecond)
```
I won't ask for a change, but if this is flaky, I'll be filing a bug against you to make a time checking interface instead of involving the filesystem in your test :)

lavalamp commented:
Sorry for delay! I just have a few more nits, then we can merge this. Are you going to make something that downloads the list?

rmmh commented:
Yes, I'm going to make a script to download the list separately. I'd like to make this a little more modular, so the Google Sheets interfacing will be a separate script.

lavalamp commented:
LGTM, please squash.
